docs/src/main/asciidoc/security-authorize-web-endpoints-reference.adoc

```diff
@@ -386,6 +386,234 @@ It is possible to use multiple expressions in the role definition.
+
+import java.security.Permission;
+import java.util.Collection;
+import java.util.Collections;
+
+@Path("/crud")
+public class CRUDResource {
+
+    @PermissionsAllowed("create") <1>
+    @PermissionsAllowed("update")
+    @POST
+    public String createOrUpdate() {
+        return "modified";
+    }
+
+    @PermissionsAllowed(value = {"create", "update"}, inclusive=true) <2>
+    @POST
+    public String createOrUpdate(Long id) {
+        return id + " modified";
+    }
+
+    @PermissionsAllowed({"see:detail", "see:all", "read"}) <3>
+    @GET
+    @Path("/id/{id}")
+    public String getItem(String id) {
+        return "item-detail-" + id;
+    }
+
+    @PermissionsAllowed(value = "list", permission = CustomPermission.class) <4>
+    @GET
+    public Collection<String> list(@QueryParam("query-options") String queryOptions) {
+        // your business logic comes here
+        return Collections.emptySet();
+    }
+
+    public static class CustomPermission extends Permission {
+
+        public CustomPermission(String name) {
+            super(name);
+        }
+
+        @Override
+        public boolean implies(Permission permission) {
+            var event = Arc.container().instance(RoutingContext.class).get(); <5>
+            var publicContent = "public-content".equals(event.request().params().get("query-options"));
+            var hasPermission = getName().equals(permission.getName());
+            return hasPermission && publicContent;
+        }
+
+        ...
+    }
+}
+----
+<1> Resource method `createOrUpdate` is only accessible by user with both `create` and `update` permissions.
+<2> By default, at least one of the permissions specified through one annotation instance is required.
+You can require all of them by setting `inclusive=true`. Both resource methods `createOrUpdate` have equal authorization requirements.
+<3> Access is granted to `getItem` if `SecurityIdentity` has either `read` permission or `see` permission and one of actions (`all`, `detail`).
+<4> You can use any `java.security.Permission` implementation of your choice.
+By default, string-based permission is performed by the `io.quarkus.security.StringPermission`.
+<5> Permissions are not beans, therefore only way to obtain bean instances is programmatically via the `Arc.container()`.
+
+CAUTION: If you plan to use the `@PermissionsAllowed` on the IO thread, review the information in xref:security-proactive-authentication-concept.adoc[Proactive Authentication].
+NOTE: The `@PermissionsAllowed` is not repeatable on class-level due to limitations of Quarkus interceptors.
+Please find well-argued explanation in the xref:cdi-reference.adoc#repeatable-interceptor-bindings[Repeatable interceptor bindings] section of the CDI reference.
+
+When the `@PermissionsAllowed` annotation is used on CDI beans, it is also possible to pass arguments, with which secured method is invoked, to custom `java.security.Permission` constructor.
```

Comment:
@sberyozkin better?

Reply:
@michalvavrik 

>When the `@PermissionsAllowed` annotation is used on CDI beans

I'm sorry if I'm a bit slow, but can you please type the examples here where `PermissionAllowed` is 1) used on the non-CDI bean 2) on CDI bean

I think you are referring to the JAX-RS endpoint being either CDI or not CDI bean, but would like to clarify

Reply:
No, I think we absolutely don't understand each other :-/ I'll explain it to you here, problem is that users won't read this PR, therefore I'll need to change docs anyway. Please don't take it personally if I will explain below in detail something that is clear, I do it for clarity :-)

- _In Quarkus all JAX-RS resources are treated as CDI beans_ see https://quarkus.io/guides/resteasy#lifecycle-of-resources, therefore I do not differentiate between JAX-RS endpoints (CDI or not)
- I use sentence _When the `@PermissionsAllowed` annotation is used on CDI beans_ and I target absolutely same meaning as here https://quarkus.io/guides/security-authorize-web-endpoints-reference#standard-security-annotations _@RolesAllowed, @DenyAll, @PermitAll on REST endpoints and CDI beans._
  - by _CDI bean_ I understand https://quarkus.io/guides/cdi-reference#lazy_by_default, for example normal scoped bean
  - by _using on_ I mean if you annotate (AKA secure) method with the `@PermissionsAllowed` that could be intercepted (e..g public etc.), you are using the `@PermissionsAllowed`
 - the situation differs from RESTEasy Reactive resources, as there, security checks are not performed by the interceptors (well, technically they are skipped) and security checks run before serialization, therefore we can't provide option to pass method arguments as there are no arguments ready at the moment when security check is done

First documentation example shows Resource with `@PermissionsAllowed`, second example (right below this sentence) shows `@PermissionsAllowed` on CDI bean.

Please feel free to suggest different wording, otherwise I will try again. Thanks!

Reply:
@sberyozkin ^^^

Reply:
I updated the documentation. Feel free to return it back to a rewrite. Thank you

Reply:
@michalvavrik No problems, though, believe it or not,  I have some idea what `using on` means, but in any case, no problems, the message I'm trying to pass to you, as a user, is that indeed, I'm aware, I thought I was aware, that CDI is always active when it comes to supporting JAX-RS resources. And I don't want to read this text as some kind of CDI specification fragment. Those `@RolesAllowed`/etc just work for me as a user on Reactive or Classic, with or without CDI scope annotations, so I'm finding these CDI highlights irrelevant, too low level, for the purpose of describing how either `RolesAllowed` and now `PermissionAllowed` work.

Now that you have updated your example, I don't think `ResourceService` is a valid JAX-RS resource, so please update it. And please think next how a user who may not be CDI savvy, can figure out when a permission with an extra property works and when it does not work (it is not clear to me why Permission with an extra parameter would not work for CRUDResource, because you can't access JAX-RS resource parameter values ? Does it mean if it is not CDI, it should be a RestEasy reactive ServerRestHandler for example that can do ?)

I'm sorry, I don't want to create a big deal here, but IMHO we need to be clear and precise in describing this feature, thanks

Reply:
OK, I see you explaining it does not work with RestEasy Reactive because the serialization is done after the security checks. In any case, this CDI bean example should show how it is accessed, is this bean injected to another JAX-RS resource or how else an HTTP call will reach it ?
As far as it is not working in Resteasy Reactive, from one hand it is not a big problem, but then from another one it is kind of a big deal, since you consider this be a cool feature and I see and agree it is but it does not work for our recommended REST stack, so can you think how to make it work with RestEasy Reactive ? Have the permission check done after the parameters have been processed ? (Can be done in another PR if agreed)

Reply:
Now that I understand your point, I can see calling it `ResourceService` was very unfortunate. I pushed updated documentation. In my eyes it's getting better, but keep feedback coming :-)

>  it does not work for our recommended REST stack, 

I agree. It is still useful as even though it doesn't work on controller/cosumer/(whatever we call where Resources are) layer, it works just fine on service layer and despise most examples in our documentation/quickstart don't have it, many bigger projects won't do application logic in the Resource (they do there validation, authN and authZ). I think some tend to secure with annotation even service layer because in bigger project it's hard to tell where the flow goes (that user permission/role is already checked).

> can you think how to make it work with RestEasy Reactive ?

TBH I based all my statements in this area on this PR https://github.com/quarkusio/quarkus/pull/19598. There are  good arguments to do security checks before anything else (including serialization), please read here https://github.com/quarkusio/quarkus/issues/16897.  IMO it is rather conceptional thing than technical - I can make it work especially with proactive authentication, but shouldn't. The situation is different on the service layer, as there, you have arguments ready (you are invoking method, no serialization mappers, JAX-RS filters etc.)

Reply:
@michalvavrik 

The examples looks good, but again, `LibraryService`, how it is accessed over HTTP ? Can you please complement that example with a JAX-RS resource showing to the users how it will be wired ?

Re Resteasy Reactive,

> I agree. It is still useful as even though it doesn't work on controller/cosumer/(whatever we call where Resources are) layer, it works just fine on service layer and despise most examples in our documentation/quickstart don't have it, many bigger projects won't do application logic in the Resource (they do there validation, authN and authZ). 

Sorry, I'm not following it. It either works with Resteasy Reactive frontend JAX-RS resources or not. Are you saying it will work even with Resteasy Reactive if the permission is applied to a CDI bean injected into the RESTeasy resource, I'm keeping guessing.

In general, IMHO, any new feature targeting JAX-RS based deployments, should be prioritized to work in the recommended JAX-RS stack which is Resteasy Reactive otherwise the new feature targeting the legacy JAX-RS stack only would have limited chance for the longer term support and thus the value of introducing such a feature becomes a bit questionable. 

This is why I believe we need to have at least a plan now how to make this new feature (Permission with an extra parameter) work in RestEasy Reactive - it can still be done before the serialization for RestEasy Reactive - but we can say that for  RestEasy Reactive only URI based parameters can be passed (path segment values, query parameters), as well as headers, which you should be able to access with Resteasy Reactive before the deserialization, either from the CDI interceptor or from `ServerRestHandler` or `ContainerRequestFilter`. It will be a reasonable restriction but will ensure RestEasy Reactive users will be able to use this new feature. (Unless it already works indirecrly as per my question above). 

What do you think ?

THanks

Reply:
Yes, I hear you, and I thank you how much time you already spent explaining this to me, but we don't seem to agree on what CDI beans are (or I misunderstood you again :-/).

> The examples looks good, but again, LibraryService, how it is accessed over HTTP ? Can you please complement that example with a JAX-RS resource showing to the users how it will be wired ?

The situation is identical to the way [RBACBean](https://github.com/quarkusio/quarkus/blob/main/integration-tests/main/src/main/java/io/quarkus/it/rest/RBACBean.java#L33) is accessed. You can see that [callingTestOnly](https://github.com/quarkusio/quarkus/blob/main/integration-tests/main/src/main/java/io/quarkus/it/rest/RBACSecuredResource.java#L96) is not secured, while you will never ever receive `callingTesterOnly` unless you have to `tester`. I consider it a proof that there is a use case for securing non-endpoints. If there is such use case, why not to allow users to pass parameters to the Permission too?

Please correct me if I am wrong, but my conclusion from you comment is:

1. Add example that calls `LibraryService`. I'll do that.

2. Have a plan for the RESTEasy Reactive endpoints, or more specifically:

> This is why I believe we need to have at least a plan now how to make this new feature (Permission with an extra parameter) work in RestEasy Reactive - it can still be done before the serialization for RestEasy Reactive - but we can say that for RestEasy Reactive only URI based parameters can be passed (path segment values, query parameters), as well as headers, which you should be able to access with Resteasy Reactive before the deserialization, either from the CDI interceptor or from ServerRestHandler or ContainerRequestFilter. It will be a reasonable restriction but will ensure RestEasy Reactive users will be able to use this new feature. (Unless it already works indirecrly as per my question above).

You are right, I think there is a lot of things that can be done safely, for example query parameters, path params and headers if they are passed to secured resource method as String or primitive type. I need to research it a bit. I'll open issue if this get merged. Right now, it doesn't make sense to open feature request (enhancement) for the feature that is not currenlty part of the Quarkus.

Reply:
@sberyozkin

Reply:
I'll wait with docs example for https://github.com/quarkusio/quarkus/pull/31687 to get merged, so that I don't waste resources.

Reply:
Done @sberyozkin , `LibraryResource` now contains resource method `updateLibrarySvc` that calls `LibraryService#updateService`. Thank you for pushing back, do you see other possibilities to improve docs?